packages/core/src/node-execution-context/node-execution-context.ts

```diff
@@ -101,6 +130,255 @@ export abstract class NodeExecutionContext implements Omit<FunctionsBase, 'getCr
+
+		// Hardcode for now for security reasons that only a single node can access
+		// all credentials
+		const fullAccess = [HTTP_REQUEST_NODE_TYPE, HTTP_REQUEST_TOOL_NODE_TYPE].includes(node.type);
+
+		let nodeCredentialDescription: INodeCredentialDescription | undefined;
+		if (!fullAccess) {
+			if (nodeType.description.credentials === undefined) {
+				throw new NodeOperationError(
+					node,
+					`Node type "${node.type}" does not have any credentials defined`,
+					{ level: 'warning' },
+				);
+			}
+
+			nodeCredentialDescription = nodeType.description.credentials.find(
+				(credentialTypeDescription) => credentialTypeDescription.name === type,
+			);
+			if (nodeCredentialDescription === undefined) {
+				throw new NodeOperationError(
+					node,
+					`Node type "${node.type}" does not have any credentials of type "${type}" defined`,
+					{ level: 'warning' },
+				);
+			}
+
+			if (
+				!NodeHelpers.displayParameter(
+					// eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+					additionalData.currentNodeParameters || node.parameters,
+					nodeCredentialDescription,
+					node,
+					node.parameters,
+				)
+			) {
+				// Credentials should not be displayed even if they would be defined
+				throw new NodeOperationError(node, 'Credentials not found');
+			}
+		}
+
+		// Check if node has any credentials defined
+		if (!fullAccess && !node.credentials?.[type]) {
+			// If none are defined check if the credentials are required or not
+
+			if (nodeCredentialDescription?.required === true) {
+				// Credentials are required so error
+				if (!node.credentials) {
+					throw new NodeOperationError(node, 'Node does not have any credentials set', {
+						level: 'warning',
+					});
+				}
+				if (!node.credentials[type]) {
+					throw new NodeOperationError(
+						node,
+						`Node does not have any credentials set for "${type}"`,
+						{
+							level: 'warning',
+						},
+					);
+				}
+			} else {
+				// Credentials are not required
+				throw new NodeOperationError(node, 'Node does not require credentials');
+			}
+		}
+
+		if (fullAccess && !node.credentials?.[type]) {
+			// Make sure that fullAccess nodes still behave like before that if they
+			// request access to credentials that are currently not set it returns undefined
+			throw new NodeOperationError(node, 'Credentials not found');
+		}
+
+		let expressionResolveValues: ICredentialsExpressionResolveValues | undefined;
+		if (connectionInputData && runExecutionData && runIndex !== undefined) {
+			expressionResolveValues = {
+				connectionInputData,
+				// eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+				itemIndex: itemIndex || 0,
+				node,
+				runExecutionData,
+				runIndex,
+				workflow,
+			} as ICredentialsExpressionResolveValues;
+		}
+
+		const nodeCredentials = node.credentials
+			? node.credentials[type]
+			: ({} as INodeCredentialsDetails);
+
+		// TODO: solve using credentials via expression
```

Comment:
Is this created in this PR or copy-pasta?

Reply:
copy-paste